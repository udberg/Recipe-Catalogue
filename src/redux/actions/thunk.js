import { fetchMeals, fetchMeal, fetchSearchMeals } from '../../API/api';
import {
  loadingMeal, errorMeal, successMeal, loadingMeals, errorMeals, successMeals,
} from './index';

const fetchMealsByCategory = (category) => async (dispatch) => {
  dispatch(loadingMeals());
  const response = fetchMeals(category);
  try {
    const data = await response;
    return dispatch(successMeals(data.meals));
  } catch (e) {
    return dispatch(errorMeals('Error while fetching data.'));
  }
};

const searchMeals = (query) => async (dispatch) => {
  dispatch(loadingMeals());
  const response = fetchSearchMeals(query);
  try {
    const data = await response;
    if (data.meals) {
      return dispatch(successMeals(data.meals));
    }
    return dispatch(errorMeals(`There is no meal with name: ${query}`));
  } catch (e) {
    return dispatch(errorMeals('Error while fetching data.'));
  }
};

const fetchMealById = (id) => async (dispatch) => {
  dispatch(loadingMeal());
  const response = fetchMeal(id);
  try {
    const data = await response;
    return dispatch(successMeal(data.meals[0]));
  } catch (e) {
    return dispatch(errorMeal('Error while fetching data.'));
  }
};

export { fetchMealsByCategory, searchMeals, fetchMealById };
