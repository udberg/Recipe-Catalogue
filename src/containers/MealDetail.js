import { useParams } from 'react-router-dom';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { connect } from 'react-redux';
import { useEffect } from 'react';
import { fetchMealById } from '../redux/actions/thunk';
import * as actionsType from '../redux/actions/actionTypes';
import card from '../styles/card.css';
import styles from '../styles/MealDetail.module.css';

const cx = classNames.bind(card);

const MealDetail = ({ meal: { status, meal, error }, dispatch }) => {
  const { id } = useParams();

  useEffect(() => {
    window.scrollTo(0, 0);
    if (id !== meal.idMeal) {
      dispatch(fetchMealById(id));
    }
  }, []);

  if (status === actionsType.LOADING_MEAL) {
    return <div>Loading ...</div>;
  }

  if (status === actionsType.ERROR_MEAL) {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  const renderIngrediants = (meal) => {
    const ingrediants = [];
    for (let index = 1; index <= 20; index += 1) {
      const ingrediant = meal[`strIngredient${index}`];
      if (ingrediant) {
        const measure = meal[`strMeasure${index}`];
        const element = (
          <li key={index}>
            <span>
              {measure}
            </span>
            <span>-</span>
            <span>
              {ingrediant}
            </span>
          </li>
        );
        ingrediants.push(element);
      }
    }
    return ingrediants;
  };

  const renderInstructions = (instructions) => {
    if (instructions) {
      const list = instructions.split(new RegExp(/\n/)).filter((item) => item.length > 1);
      return (
        <ul className={cx(styles.items)}>
          {/* eslint-disable-next-line */}
          {list.map((item, index) => <li key={index}>{item}</li>)}
        </ul>
      );
    }
    return instructions;
  };

  return (
    <div>
      <div className={cx('card-detail')}>
        <img className={cx('card-img')} src={meal.strMealThumb} alt={meal.strMeal} />
        <div>
          <h3 className={cx('card-title')}>{meal.strMeal}</h3>
          <div className={cx(styles.ingredients)}>
            <h4 className={cx(styles.heading)}>Ingredients</h4>
            <ul className={cx(styles.listItem)}>
              {renderIngrediants(meal)}
            </ul>
          </div>
          <div className={cx(styles.instructions)}>
            <h4 className={cx(styles.heading)}>Instructions</h4>
            <div data-testid="meal-description" className={cx(styles.instructions)}>
              {renderInstructions(meal.strInstructions)}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

MealDetail.propTypes = {
  meal: PropTypes.shape({
    status: PropTypes.string.isRequired,
    error: PropTypes.string,
    meal: PropTypes.objectOf(PropTypes.string),
  }).isRequired,
  dispatch: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  meal: state.meal,
});

export default connect(mapStateToProps)(MealDetail);
