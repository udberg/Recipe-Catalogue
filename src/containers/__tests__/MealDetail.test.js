import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import ConfigureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MealDetail from '../MealDetail';

const mockStore = ConfigureMockStore([thunk]);
window.scrollTo = jest.fn();
jest.mock('../../API/api');

test('should test for empty meal', () => {
  const store = mockStore({
    meal: {
      meal: {},
      status: '',
      error: null,
    },
  });
  const { getByTestId } = render(
    <Provider store={store}>
      <MemoryRouter>
        <MealDetail />
      </MemoryRouter>
    </Provider>,
  );

  expect(getByTestId('meal-description').textContent).toMatch('');
});

test('should test meal after action is dispatched', () => {
  const store = mockStore({
    meal: {
      meal: {
        strMeal: 'Testing meal',
        strInstructions: 'This is an instructions for a meal.',
      },
      status: '',
      error: null,
    },
  });

  const { container, getByTestId } = render(
    <Provider store={store}>
      <MemoryRouter>
        <MealDetail />
      </MemoryRouter>
    </Provider>,
  );

  expect(getByTestId('meal-description').textContent).toMatch('This is an instructions for a meal.');
  expect(container.firstChild).toMatchSnapshot();
});
