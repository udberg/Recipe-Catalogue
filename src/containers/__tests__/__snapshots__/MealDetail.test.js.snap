// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test meal after action is dispatched 1`] = `
<div>
  <div
    class="card-detail"
  >
    <img
      alt="Testing meal"
      class="card-img"
    />
    <div>
      <h3
        class="card-title"
      >
        Testing meal
      </h3>
      <div
        class="ingredients"
      >
        <h4
          class="heading"
        >
          Ingredients
        </h4>
        <ul
          class="listItem"
        />
      </div>
      <div
        class="instructions"
      >
        <h4
          class="heading"
        >
          Instructions
        </h4>
        <div
          class="instructions"
          data-testid="meal-description"
        >
          <ul
            class="items"
          >
            <li>
              This is an instructions for a meal.
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
